Optimise '(caller)[ ...]' into just OP_CALLER with private flags

richardleach opened the pull request:
A subroutine can obtain just the package of its caller in a couple of ways. Both seem somewhat common.

 * `caller` - in scaler context, as in `my $x = caller;`
 * `(caller)[0]`, as in `my $x = (caller)[0];`

In the first, `caller` finds the package name, sticks it in a new SV, and puts that (or `undef`) on the stack:

    <0> caller[t2] s

In the second, `caller` (a) finds the package name, filename, and line (b) creates three new SVs to hold them all (c) puts those SVs on the stack (d) does a list slice to leave just the package SV on the stack.

    7        <2> lslice sK/2 ->8
    -           <1> ex-list lK ->5
    3              <0> pushmark s ->4
    4              <$> const[IV 0] s ->5
    -           <1> ex-list lK ->7
    5              <0> pushmark s ->6
    6              <0> caller[t2] l ->7

This PR originally checked for the second case inside `Perl_newSLICEOP` and instead of constructing a `lslice` OP, returned just the `caller` OP with scalar context applied.

_Update July 2026:_

Following feedback in this PR, the unused 7 bits of `op_private` are now used to encode the following subscripts, which seem to be the most commonly occuring in core and on CPAN.
* 0 - package
* 1 - filename
* 2 - line
* 3 - subroutine
* 8 - hints
* 9 - bit mask
* 10 - hint hash

The optimization supports a single subscript, as in the original PR, or an ascending run.

So these will be optimized:
* `(caller)[3]`
* `(caller)[1,2,3]`
* `(caller)[0,3,9]` - this one not seen in the wild!
* `(caller)[0,1,2,3,8,9,10]` - ditto!

But these will not:
* `(caller)[4]`
* `(caller)[2,1,3]`
* `(caller)[2,3,4]`

If a compelling case to support subscripts 4-7 presents itself, `OP_CALLER` will probably have to become an `UNOP_AUX` in order to have sufficient bits.

Arbitrary subscript ordering doesn't seem common enough to be worth the effort of implementing, but if a compelling argument was made, that would probably also need `OP_CALLER` to become an `UNOP_AUX`.

---------------------------------------------------------------------------------
* This set of changes will require a perldelta entry.

bulk88 commented:
https://grep.metacpan.org/search?size=20&_bb=86025425&q=%5C%28caller%5C%29%5C%5B&qft=*.pm%2C+*.t&qd=&qifl=

Should've been done 30 years ago. Most PP devs think `(caller())[\w]` is constant folded as if Perl is identical to C++. The truth is list context PP `context()` is like writing a 100MB core dump to a SSD everytime you execute it. `caller()` shouldve never ever have become Perl best practices/cargo culted. There is nothing wrong with `caller`'s public PP API IMO, but b/c of its horrible runtime internal implementation, it shouldve been sent to the landfill, by `use strict;` on day 1 of `use strict;`, just like this Perl 5 code was sent to the landfill by `use strict;`

```
C:\Users\Owner>perl -e" push( @a, THISS); push( @a, ISS); push( @a, PERL); $, =' '; print @a;"
THISS ISS PERL
C:\Users\Owner>
```

Rich, would you pretty please be able to tackle adding OP tree compile time next gen `G_VOID` propagation to the other 12-15 list context retval indexes/retval SV*s created by `pp_caller`?

richardleach commented:
Hmmm, if I can co-opt `op_private` then it's likely possible to cover individual elements 1,2,3 and in-order slices like `[1,2]`, which seem to make up the majority of usage on CPAN. I'll have a go.

bulk88 commented:
> Hmmm, if I can co-opt `op_private` then it's likely possible to cover individual elements 1,2,3 and in-order slices like `[1,2]`, which seem to make up the majority of usage on CPAN. I'll have a go.

pp_caller() has 3 return prototypes,

-1 SV*
-3 SV*s
-10 SV*s

All that is needed is a room to put a U16 variable in OP_CALLER's OP struct, or 10 unused bits somewhere. 10 bits will allow a fast and easy pattern of
```
if(op->opaque & 0x4) {
    PUSH(sumsv);
}
if(op->opaque & 0x8) {
    PUSH(sumsv);
}
if(op->opaque & 0x10) {
    PUSH(sumsv);
}
```
inside pp_caller.

My thought 5 years ago was to add a 2nd integer argument, allowing the end use to pick 1 of 10 elements to return, but I didn't like my proposal, since it would only help new code, and heavily policed and evangelized new code.
And "heavily policed and evangelized new code"   really means @bulk88 is making PRs to various CPAN modules and demanding those authors make a new CPAN release tar.gz for @bulk88's brand new perl 5 grammer enhancement authored by @bulk88.

My wishlist quickly changed to, the existing production in the field deployed perl code needs to be left un-touched, the correct fix would be from the P5P side, from the 'yylex'/'ck_op_()' side by analyzing the scalar or list context `=` operator and the target lvalue somehow, and eventually see if it was assigned to a `$`/`@` lvalue or to an anonymous array, array ref, whatever this is `my $x = (caller)[0];`.

```
7        <2> lslice sK/2 ->8
-           <1> ex-list lK ->5
3              <0> pushmark s ->4
4              <$> const[IV 0] s ->5
-           <1> ex-list lK ->7
```
I don't know how `ex-list` and `lslice` pp_foo() funcs work from the top of my head, I very rarely or never see them flash by while holding F11 or breakpointing each cycle of `runops_std()`. But back to my point, the wish list is to extract that array dereference const literal integer from `[]` operator's OP, and stick it into `caller()` operator's OP.

richardleach commented:
> _-10 SV_s

I didn't do an exhaustive grep of CPAN on the train this morning, but it did seem like handling only the first 4 SVs would cover the vast majority of actually-encounted slice cases. Hence abuse of `op_private` seemed like it might be enough.

bulk88 commented:
> > _-10 SV_s
> 
> I didn't do an exhaustive grep of CPAN on the train this morning, but it did seem like handling only the first 4 SVs would cover the vast majority of actually-encounted slice cases. Hence abuse of `op_private` seemed like it might be enough.

values 0-9 aka 10 SV*s fit in 4 bits. Picking 1, 2 , 3 or more non linear SV*s out of 10, requires 10 bits of space. 

IDK how to read all the lines below, but my eyes see worst case ever 3 bits free, at minimum case a U8 available, getting creative steal some bits from `U8 op_flags`, after that add another type code inside  `    PERL_BITFIELD16 op_type:9;`, after that steal `    PADOFFSET	op_targ;`, im not sure a caller in list context, can have a `TARG` since isn't `TARG` only for `caller()` in scalar context  , and then `TARG` is the lvalue on the left side `caller()`? Im guessing but not checking the src code to verify, that `caller()` always pulls its 1 and only incoming arg, which is an `IV`/`U32` of how many PP frames to go backwards, from the PL stack, not from `caller()` OP's `TARG`, correct me if I am wrong.

```
caller		caller			ck_fun		t%	S?
# baseop/unop - %
#define BASEOP				\
    OP*		op_next;		\
    OP*		op_sibparent;		\
    OP*		(*op_ppaddr)(pTHX);	\
    PADOFFSET	op_targ;		\
    PERL_BITFIELD16 op_type:9;		\
    PERL_BITFIELD16 op_opt:1;		\
    PERL_BITFIELD16 op_slabbed:1;	\
    PERL_BITFIELD16 op_savefree:1;	\
    PERL_BITFIELD16 op_static:1;	\
    PERL_BITFIELD16 op_folded:1;	\
    PERL_BITFIELD16 op_moresib:1;       \
    PERL_BITFIELD16 op_spare:1;		\
    U8		op_flags;		\
    U8		op_private;
#endif
    /* CALLER     */ (OPpARG4_MASK|OPpOFFBYONE),
#define OPpARG4_MASK            0x0f
#define OPpOFFBYONE             0x80
```

There is also the sneaky solution of `pp_caller()` at runloop time doing a sneaky deref into a const folded/disabled but not defragmented OP *, and learning the const literal integer from a different OP* struct than its own OP* struct. Certain other pp_*() funcs do this design pattern already.

richardleach committed:
pp_caller: use OPf_KIDS rather than MAXARG to free up bits

richardleach committed:
pp_caller: use OPf_KIDS rather than MAXARG to free up bits

richardleach committed:
pp_caller: do the EXTEND check once and early

richardleach committed:
pp_caller: do the EXTEND check once and early

richardleach committed:
pp_caller: minor refactoring ahead of PP unwrapping

richardleach committed:
pp_caller: minor refactoring ahead of PP unwrapping

richardleach committed:
pp_caller - RC stack unwrapped

richardleach committed:
pp_caller - RC stack unwrapped

richardleach committed:
OP_CALLER: handle common list slices of return values internally

It's common for `caller()` to be called in list context and immediately
sliced to get a particular return value or a subset thereof. For example:

    my $sub = (caller)[3];

or

    my ($file, $line) = (caller 1)[1,2];

Historically, this meant producing the full set of caller's return values
(3 if no EXPR was provided, 11 if it was), often requiring the creation
and population of new SVs, only for some/most of them to be immediately
discarded by the list slice.

This commit makes use of available `op_private` bits to encode what seem
like the most commonly used (in core and on CPAN) subscripts inside
an OP_CALLER instance that will be immediately sliced. When this occurs,
the actual OP_LSLICE OP and it's OP_CONST children are freed.

Those common & supported subscripts are:
*  0 (package)
*  1 (filename)
*  2 (line)
*  3 (subroutine)
*  8 (hints)
*  9 (bitmask)
* 10 (hinthash)

Ascending runs of these are also supported. So these will be optimized:

* (caller)[3]
* (caller)[1,2,3]
* (caller)[0,3,9] - this one not seen in the wild!

But these will not:

* (caller)[4]
* (caller)[2,1,3]
* (caller)[2,3,4]

At runtime, `pp_caller` checks if any of the relevant bits are set and,
if so, generates and emits only the requested values. (If multiple
subscripts are requested but the OP has scalar context applied, the
set of `op_private` bits will be scanned and only the highest kept.)

This turns `my $sub = caller)[3];` from:

    7        <2> lslice sK/2 ->8
    -           <1> ex-list lK ->5
    3              <0> pushmark s ->4
    4              <$> const[IV 0] s ->5
    -           <1> ex-list lK ->7
    5              <0> pushmark s ->6
    6              <0> caller[t2] l ->7

into:

    <0> caller[t2] s

And `my ($file, $line) = (caller 1)[1,2];` from:

    c     <2> aassign[t4] vKS ->d
    -        <1> ex-list lK ->b
    3           <0> pushmark s ->4
    a           <2> lslice lK/2 ->b
    -              <1> ex-list lK ->7
    4                 <0> pushmark s ->5
    5                 <$> const[IV 1] s ->6
    6                 <$> const[IV 2] s ->7
    -              <1> ex-list lK ->a
    7                 <0> pushmark s ->8
    9                 <1> caller[t3] lK/1 ->a
    8                    <$> const[IV 1] s ->9
    -        <1> ex-list lKPRM* ->c
    b           <0> padrange[$file:1,2; $line:1,2] RM/LVINTRO,range=2 ->c

into:

    7     <2> aassign[t4] vKS ->8
    -        <1> ex-list lK ->6
    3           <0> pushmark s ->4
    5           <1> caller[t3] lK/LINE,FILE ->6
    4              <$> const(IV 1) s ->5
    -        <1> ex-list lKPRM* ->7
    6           <0> padrange[$file:1,2; $line:1,2] RM/LVINTRO,range=2 ->7

richardleach committed:
Perldelta entry for optimized caller() list slices [GH #23369]

tonycoz commented on op.c:
```diff
@@ -8989,6 +9008,77 @@ Perl_dofile(pTHX_ OP *term, I32 force_builtin)
+     *
+     * At present, only subscripts 0 - 3 are supported, and these
+     * seem to predominate on CPAN. However, [9] is involved in building
+     * and [8], [10], and [8,9,10] are also present
```
Out of date comment?

richardleach replied:
Yes, thanks. Will update.

tonycoz commented on op.c:
```diff
@@ -2069,6 +2069,25 @@ Perl_scalar(pTHX_ OP *o)
+                            }
+                            break;
+                        }
+                    }
```
Maybe:
```
                    o->op_private = (1U << msbit_pos32(caller_private))
                        | (o->op_private & OPpOFFBYONE);
```

I half expected gcc to recognize the loop and use the bsr instruction, but alas it didn't.

clang unrolled the loop which made me cry.

richardleach replied:
Thanks, that's much nicer than the loop.

tonycoz commented on pp_ctl.c:
```diff
@@ -2392,6 +2392,103 @@ PP(pp_caller)
+                return NORMAL;
+        }
+
+        if (subscripts & OPpCALLER_SUB) {
```
Probably need:
```
if (!has_arg)
    return NORMAL;
```
here, eg:

```
# this PR:
tony@venus:.../git/perl6$ ./perl -Ilib -E 'sub f { scalar +(caller)[0,2,10] } say f()'
HASH(0x558c25562618)
# 5.40.1 debian:
tony@venus:.../git/perl6$ perl -E 'sub f { scalar +(caller)[0,2,10] } say f()'
 
tony@venus:.../git/perl6$
```

tonycoz replied:
The scalar was left over in my command history, without:
```
tony@venus:.../git/perl6$ ./perl -Ilib -E 'sub f { (caller)[0,2,10] } say f()'
main1HASH(0x55e9629cf660)
tony@venus:.../git/perl6$ perl -E 'sub f { (caller)[0,2,10] } say f()'
main1
tony@venus:.../git/perl6$
```

tonycoz replied:
Actually, my suggested fix is wrong, if I adjust my test case:
```
tony@venus:.../git/perl6$ ./perl -Ilib -E 'sub f { (caller)[0,2,10] } say join ":", f()'
main:1:HASH(0x564b09528660)
tony@venus:.../git/perl6$ perl -E 'sub f { (caller)[0,2,10] } say join ":", f()'
main:1:
tony@venus:.../git/perl6$
```
My suggestion would eliminate the `undef` returned for the missing element 10.

richardleach replied:
Thanks. I'll add an `if (!has_arg)` check for all subscripts above `2`.

tonycoz commented on pp_ctl.c:
```diff
@@ -2395,82 +2394,84 @@ PP_wrapped(pp_caller, ((PL_op->op_flags & OPf_KIDS) ? 1 : 0), 0)
         else {
-            PUSHs(newSVpvs_flags("(unknown)", SVs_TEMP));
-            PUSHs(boolSV(CxHASARGS(cx)));
+            rpp_push_1_norc( newSVpvs_flags("(unknown)", 0));
```
Don't need the `_flags` variant anymore.

tonycoz replied:
Some other places too.

richardleach replied:
Tweaked.

tonycoz commented on pp_ctl.c:
```diff
@@ -2395,82 +2394,84 @@ PP_wrapped(pp_caller, ((PL_op->op_flags & OPf_KIDS) ? 1 : 0), 0)
             else {
                 /* I think this is will always be "", but be sure */
-                PUSHs(sv_mortalcopy_flags(cur_text, SV_GMAGIC|SV_NOSTEAL));
+                rpp_push_1_norc(newSVsv_flags(cur_text, SV_GMAGIC|SV_NOSTEAL));
```
Don't need the newSVsv `_flags` variant here, GMAGIC|NOSTEAL is the default.

richardleach replied:
True, I just like being able to see what the flags are without having to go and check. 😁

tonycoz commented on pp_ctl.c:
```diff
@@ -2319,14 +2319,16 @@ PP_wrapped(pp_caller, ((PL_op->op_flags & OPf_KIDS) ? 1 : 0), 0)
-      else (void)POPs;
+        if (PL_stack_sp[0]) {
+            has_arg = true;
+            count += (IV)SvIVx(PL_stack_sp[0]);
```
Why the IV cast?

Why the SvIVx() variant instead of just SvIV()?  It's exactly the same as SvIV() due to Karl's effort to avoid macro parameter multiple evaluation.

richardleach replied:
Probably mindless copy-pasta, I'll change it to `SvIV()`.

tonycoz commented on pp_ctl.c:
```diff
@@ -2392,6 +2392,103 @@ PP(pp_caller)
+     * return values here, rather than complicating the code flow
+     * of regular pp_caller. */
+    if (UNLIKELY(PL_op->op_private && (PL_op->op_private != OPpOFFBYONE))) {
+        assert(cx);
```
Not sure why this assert is here.

Both sides of the conditional need cx non-NULL, and the check after caller_cx() seems like it's covered anyway.

richardleach replied:
Left over after a mid-development refactor, I think. Will remove it.

tonycoz commented on pp_ctl.c:
```diff
@@ -2392,6 +2392,103 @@ PP(pp_caller)
+                }
+            } else {
+                rpp_push_1_norc( newSVpvs_flags("(eval)", 0) );
+            }
```
The duplicated code makes me sad.

In particular for SUB, BITS, maybe PKG.

Rather than complicating this code they could just call static inline functions defined immediately above pp_caller.

richardleach commented:
Thanks for the review comments, @tonycoz. I've pushed a few updates, but still have to look at deduplicating some code.

tonycoz commented on op.c:
```diff
@@ -2080,6 +2080,27 @@ Perl_scalar(pTHX_ OP *o)
+//                            }
+//                            break;
+//                        }
+//                    }
```
Leftovers

richardleach replied:
Derp, removing now.

tonycoz commented on pp_ctl.c:
```diff
@@ -2379,108 +2466,147 @@ PP_wrapped(pp_caller, MAXARG, 0)
-            dTARGET;
-            sv_sethek(TARG, stash_hek);
-            PUSHs(TARG);
+            switch(subscripts) {
```
GIMME is dynamic, the op can be in scalar context without scalar() having been called on it:
```
tony@venus:.../git/perl6$ perl -E 'sub f { (caller 0)[0, 1] } say scalar f()'
-e
tony@venus:.../git/perl6$ ./perl -Ilib -E 'sub f { (caller 0)[0, 1] } say scalar f()'
 
tony@venus:.../git/perl6$
```

richardleach replied:
Ohhhh, yup. That branch now pulls the msb.

tonycoz reviewed (approved):
Though hopefully some of it can be squashed.

Given the discussion on IRC it might be worth moving the bitcount loop to an inline function than we can adapt later, but that can be a follow-up
